packages/grid_client/tests/modules/applications/casperlabs.test.ts

```diff
@@ -0,0 +1,267 @@
+     Test Cases: TC2683 - Applications: Deploy Casperlabs
+     Scenario:
+        - Generate Test Data/casperlabs Config/Gateway Config.
+        - Select a Node To Deploy the casperlabs on.
+        - Select a Gateway Node To Deploy the gateway on.
+        - Deploy the casperlabs solution.
+        - Assert that the generated data matches
+          the deployment details.
+        - Pass the IP of the Created casperlabs to the Gateway
+          Config.
+        - Deploy the Gateway.
+        - Assert that the generated data matches
+          the deployment details.
+        - Assert that the Gateway points at the IP
+          of the created casperlabs.
+        - Assert that the returned domain is working
+          and returns correct data.
+    **********************************************/
+
+  //Test Data
+  const name = "gw" + generateString(10).toLowerCase();
+  const tlsPassthrough = false;
+  let cpu = generateInt(1, 4);
+  let memory = generateInt(1, 6);
+  let rootfsSize = generateInt(0, 2);
+  let diskSize = generateInt(10, 50);
+  const networkName = generateString(15);
+  const vmName = generateString(15);
+  const diskName = generateString(15);
+  const mountPoint = "/data";
+  const publicIp = false;
+  const ipRangeClassA = "10." + generateInt(1, 255) + ".0.0/16";
+  const ipRangeClassB = "172." + generateInt(16, 31) + ".0.0/16";
+  const ipRangeClassC = "192.168.0.0/16";
+  const ipRange = randomChoice([ipRangeClassA, ipRangeClassB, ipRangeClassC]);
+  const metadata = "{'deploymentType': 'casperlabs'}";
+  const description = "test deploying Casperlabs via ts grid3 client";
+
+  //GatewayNode Selection
+  const gatewayNodes = await gridClient.capacity.filterNodes({
+    gateway: true,
+    farmId: 1,
+    availableFor: await gridClient.twins.get_my_twin_id(),
+  } as FilterOptions);
+  if (gatewayNodes.length == 0) throw new Error("no nodes available to complete this test");
+  const GatewayNode = gatewayNodes[generateInt(0, gatewayNodes.length - 1)];
+
+  //Node Selection
+  let nodes;
+  try {
+    nodes = await gridClient.capacity.filterNodes({
+      cru: cpu,
+      mru: memory,
+      sru: rootfsSize + diskSize,
+      farmId: 1,
+      availableFor: await gridClient.twins.get_my_twin_id(),
+    } as FilterOptions);
+  } catch (error) {
+    //Log the resources that were not found.
+    log("A Node was not found with the generated resources." + error);
+    log("Regenerating test data with lower resources....");
+
+    //Generate lower resources.
+    cpu = generateInt(1, cpu);
+    memory = generateInt(1, memory);
+    rootfsSize = generateInt(0, rootfsSize);
+    diskSize = generateInt(5, diskSize);
+
+    //Search for another node with lower resources.
+    nodes = await gridClient.capacity.filterNodes({
+      cru: cpu,
+      mru: memory,
+      sru: rootfsSize + diskSize,
+      farmId: 1,
+      availableFor: await gridClient.twins.get_my_twin_id(),
+    } as FilterOptions);
```

Comment:
I think we should stop the test if no node w valid resources is found.

Reply:
Yes, I throw an error; it should stop the test. `throw new Error("no nodes available to complete this test");`.